### ISSUE TITLE
When I scan the C# project with the docker container, there seems to be an error, integer overflow, and no vulnerabilities scanned

### ISSUE THREAD

**Issue by chenkehao1998** [microsoft/infersharp#152]
```
[root@MiWiFi-R3600-srv Miniblog.Core]# docker run -v /home/Jonas/Project/Miniblog.Core:/infersharp/binary_path --rm mcr.microsoft.com/infersharp:v1.3 /bin/bash -c "./run_infersharp.sh binary_path; cp infer-out/report.txt /infersharp/binary_path/report.txt"
Processing {binary_path}
Copying binaries to a staging folder...

Code translation started...
Translation stage 1/3: Loading binaries.
Translation stage 2/3: Computing type environment.
Translation stage 3/3: Computing control-flow graph.

Coverage Statistics:

Method successfully translated: 0 (-2147483648%)
Method partially translated: 0 (-2147483648%)
Instructions translated: 0 (-2147483648%)
Instructions skipped: 0 (-2147483648%)
======================================

Code translation completed. Analyzing...


  No issues found
```

```
[root@MiWiFi-R3600-srv Miniblog.Core]# docker run -v /home/Jonas/Project/Miniblog.Core:/infersharp/Examples -it mcr.microsoft.com/infersharp:v1.3
root@9facf129f394:/infersharp# ./run_infersharp.sh Examples
Processing {Examples}
Copying binaries to a staging folder...

Code translation started...
Translation stage 1/3: Loading binaries.
Translation stage 2/3: Computing type environment.
Translation stage 3/3: Computing control-flow graph.

Coverage Statistics:

Method successfully translated: 0 (-2147483648%)
Method partially translated: 0 (-2147483648%)
Instructions translated: 0 (-2147483648%)
Instructions skipped: 0 (-2147483648%)
======================================

Code translation completed. Analyzing...


  No issues found
```

**Comment by Jonas-sci** [comment]
I also find this problem 
======================
```
[root@MiWiFi-R3600-srv Project]# docker run -v /home/Jonas/Project/OrchardCore/:/infersharp/Examples -it mcr.microsoft.com/infersharp:v1.3 
root@ffd2c246885c:/infersharp# ./run_infersharp.sh Examples/
Processing {Examples/}
Copying binaries to a staging folder...

Code translation started...
Translation stage 1/3: Loading binaries.
Translation stage 2/3: Computing type environment.
Translation stage 3/3: Computing control-flow graph.

Coverage Statistics:

Method successfully translated: 0 (-2147483648%)
Method partially translated: 0 (-2147483648%)
Instructions translated: 0 (-2147483648%)
Instructions skipped: 0 (-2147483648%)
======================================

Code translation completed. Analyzing...


  No issues found  
```

**Comment by chenkehao1998** [comment]
When I scan the example , this situation will not occured .

**Comment by Jonas-sci** [comment]
> When I scan the example , this situation will not occured .

when I scan [example](https://github.com/microsoft/infersharp/blob/main/RUNNING_IN_DOCKER.md) ,I can get this result
```
[root@MiWiFi-R3600-srv Project]# docker run -it mcr.microsoft.com/infersharp:v1.3
root@5972e11cfb91:/infersharp# ./run_infersharp.sh Examples/ 
Processing {Examples/}
Copying binaries to a staging folder...

Code translation started...
Translation stage 1/3: Loading binaries.
Translation stage 2/3: Computing type environment.
Translation stage 3/3: Computing control-flow graph.

Coverage Statistics:

Method successfully translated: 11 (100%)
Method partially translated: 0 (0%)
Instructions translated: 143 (100%)
Instructions skipped: 0 (0%)
======================================

Code translation completed. Analyzing...


/Examples/Examples/Program.cs:53: error: Dotnet Resource Leak
  Leaked resource sw of type System.IO.StreamWriter in method "Void Program.ResourceLeakIntraproceduralBad()".

/Examples/Examples/Program.cs:78: error: Dotnet Resource Leak
  Leaked resource stream of type System.IO.StreamWriter in method "Void Program.ResourceLeakInterproceduralBad()".

/Examples/Examples/Program.cs:124: error: Dotnet Resource Leak
  Leaked resource p.Examples.Program.SWGlobal of type System.IO.StreamWriter, Leaked resource p.Examples.Program.SRGlobal of type System.IO.StreamReader in method "Void MainClass.Main(String[])".

/Examples/Examples/Program.cs:128: error: Null Dereference
  ``String Program.NullDeReferenceBad()`` could be null (from the call to `Program.NullDeReferenceBad()` on line 128) and is dereferenced.


Found 4 issues
                  Issue Type(ISSUED_TYPE_ID): #
  Dotnet Resource Leak(DOTNET_RESOURCE_LEAK): 3
       Null Dereference(NULLPTR_DEREFERENCE): 1

```

they can scan the vulnerability.and percent is correct

**Comment by chenkehao1998** [comment]
> @chenkehao1998, @Jonas-sci, thanks for reporting this. I should have made the instructions clearer. I will make an update to the doc.
> 
> There are [two modes running Docker](https://github.com/microsoft/infersharp/blob/main/RUNNING_IN_DOCKER.md) - direct and interactive. You are using the direct mode. In that case, `path_to_binary_folder` should be a Windows path. For example, `C:\project1\bin`. If the binaries are already at a Linux path, you can use the interactive mode - `./run_infersharp.sh Examples/` where `Examples/` is the path that is relative to the script.

Thanks for your reply!!!
But I have used the interactive mode , the result is still not correct .

I used the command like this.
```
docker run 
-v /home/Jonas/Project/Miniblog.Core:/infersharp/Examples 
-it mcr.microsoft.com/infersharp:v1.3
```
And then ,
```
./run_infersharp.sh Examples
```
the file in directory  `Examples` is my C# project files now.

but the result is 
```
Processing {Examples}
Copying binaries to a staging folder...

Code translation started...
Translation stage 1/3: Loading binaries.
Translation stage 2/3: Computing type environment.
Translation stage 3/3: Computing control-flow graph.

Coverage Statistics:

Method successfully translated: 0 (-2147483648%)
Method partially translated: 0 (-2147483648%)
Instructions translated: 0 (-2147483648%)
Instructions skipped: 0 (-2147483648%)
======================================

Code translation completed. Analyzing...


  No issues found
```

**Comment by xinshiMSFT** [comment]
The `Example` folder comes with the Docker image. It provides some example binaries that you can try. To run your own binaries, assuming you are using Windows Command Prompt, the exact steps for the interactive mode are:

1. Run `docker run -it mcr.microsoft.com/infersharp:v1.3`
2. Open another Command Prompt, run `docker ps -a` to find out the container ID. Then run `docker cp <your project folder path on Windows> <container_id>:infersharp/<your project folder name>`
3. Go back to the other Command Prompt from step 1 and run `./run_infersharp.sh <your project folder name>`

**Comment by m-sedl** [comment]
Hi!
I also found this bug. I have run InferSharp 1.4 and InferSharp 1.3 via docker container on two systems: `macos 12.0.1` and `ubuntu 22.04.01`.

I did different launches, I will describe each one separately.

1) InferSharp 1.4 via docker interactively on macos. Ran `./run_inferharp Examples`. Everything works perfectly.

2) InferSharp 1.4 via docker interactively on ubuntu . Ran `./run_inferharp Examples`. I get a known bug:
  ```
  Code translation started...
  Translation stage 1/3: Loading binaries.
  Translation stage 2/3: Computing type environment.
  Translation stage 3/3: Computing control-flow graph.
  
  Coverage Statistics:
  
  Method successfully translated: 0 (-2147483648%)
  Method partially translated: 0 (-2147483648%)
  Instructions translated: 0 (-2147483648%)
  Instructions skipped: 0 (-2147483648%)
  ```

3) InferSharp 1.4 via docker in direct mode on macos. I built a project with [this file](https://gist.github.com/m-sedl/8ef88191ad73cfc172a5eff2f26c9ee0) using `dotnet 6.0.402` on macos . Everything works perfectly.

4) InferSharp 1.4 via docker in direct mode on ubuntu . I built the same project but using `dotnet 6.0.110` in ubutnu . The bug is reproducing.

5) InferSharp 1.4 via docker in direct mode on ubuntu . I used the results of building a test project from macos. Everything works perfectly. I compiled on macos for different runtimes (linux-x64, osx.12-x64). InferSharp 1.4 was able to analyze all the assembly variants made in macos.

6) InferSharp 1.3 works correctly through docker (in both modes) on both ubuntu and macos.

I found that the difference is in how assemblies made on macos and ubuntu are loaded. In the second case, the dll's cannot be loaded by [this method](https://github.com/microsoft/infersharp/blob/main/Cilsil/Services/DecompilationService.cs#L50). A `SymbolsNotMatchingException` is thrown [in the Mono.cecil library](https://github.com/jbevain/cecil/blob/d820a5beca4104364178c61e1c59718fb228589e/Mono.Cecil/ModuleDefinition.cs#L1080). Because of this, the assembly is considered broken and cecil cannot work with it correctly, so none of the methods are loaded.

It is possible that since the release of version 1.3 you somehow affected the logic of loading assemblies, or one of your docker image dependencies was updated in such a way that the loading of dlls built on linux broke.

I will try to debug this problem, but I consider it necessary to tell you everything that I managed to find.

**Comment by xinshiMSFT** [comment]
@m-sedl, thank you so much for providing the detailed repro steps!

We recently updated the script to speed up the binary copying process (https://github.com/microsoft/infersharp/commit/bc6fb62d01115a6edc81ad46e15d75934be1fa68). Instead of copying everything from a user-specified folder, the script only copies the .dll and pdb files and flattens the structure.

.NET produces `Examples.dll` under the `Examples/ref` folder after `dotnet publish`. However, it is not the matching .dll file with `Examples/Examples.pdb`. On Linux, the script probably copied `Examples/ref/Examples.dll` and `Examples/Examples.pdb` first, then ignored `Examples/Examples.dll`. It is probably why you are seeing a `SymbolNotMatchingException`.

I don't have a Mac, but I am guessing the reason you are not seeing it on MacOS is that the copying behavior is different. It probably copied the matching `Examples/Examples.dll` and `Examples/Example.pdb` first, then ignored `Examples/ref/Examples.dll`.

I made a PR https://github.com/microsoft/infersharp/pull/179 so that the copying behavior becomes consistent across different OSs. Would you mind updating `run_infersharp.sh` in your local docker container and try again?

**Comment by xinshiMSFT** [comment]
@m-sedl, you are probably right. It looks like the crash comes from the Infer backend. But just to be sure, can you try reverting `run_infersharp.sh` to [this commit](https://github.com/microsoft/infersharp/blob/19bf547f15b07e2227f97c49ba7520ed7370e2b9/run_infersharp.sh) and run the analysis with the 1.4 bits again?

Please note that we currently do not support MacOS officially, but we will see what we can do to unblock you. If you are still seeing the same issue on MacOS, do you mind sharing the binaries with us, or pointing us to the commit/version of NLog on GitHub so that we can investigate on our side?

**Comment by m-sedl** [comment]
@xinshiMSFT, reverting `run_infersharp.sh` helped. 1.4 version with reverted script worked successfully on NLog project on macos

**Comment by xinshiMSFT** [comment]
@m-sedl, thank you! It is probably due to the copy logic then. Good to know that at least you are not blocked.

Can you please share the NLog binaries on your machine with us so that we are sure we are looking at the same binaries? Or if the size is too large or it is not convenient, can you point us to the commit/version of NLog you are using so we can build it ourselves?

**Comment by m-sedl** [comment]
Извини за то, что сразу не сбросил ссылку. 
commit: `fe7ed742c28a179ec96a2c26fc6fc8d60078780b`
бинарные файлы, собранные на macos: https://disk.yandex.ru/d/Ns2-3nSvFXcxDA

**Comment by m-sedl** [comment]
@xi-liu-ds,
Hmm. Very strange, but I can't reproduce the error. Now everything is working properly. I'm sorry for the inconvenience.

My docker desktop version is 4.12.0
My parameters:
<img width="1382" alt="image" src="https://user-images.githubusercontent.com/28401551/196800454-eab621af-0783-49f7-b5b4-cba02ebb3796.png">

**Comment by xi-liu-ds** [comment]
Closing this issue. Please feel free to open a new issue if runs into this problem again.

**Comment by BlairMcClelland** [comment]
I had the same error when using Windows 11/Docker, it turns out my .wslconfig looked like this
```
[wsl2]
memory=2GB
swap=0
```
After changing it to below all is well

```
[wsl2]
memory=8GB
```
